### PR TITLE
Use ECMAScript modules (ESM) nearly everywhere

### PR DESCRIPTION
**Pull request by mxdvl**
## What does this change?

Unify (nearly) all our modules to be ECMAScript modules

- CommonJS modules get the explicit `.cjs` extension
- Jest tests uses the [unstable ESM mocking API](https://jestjs.io/docs/ecmascript-modules#module-mocking-in-esm), which requires mocked modules to be imported after the mock using dynamic imports.

## Why?

Consitency, since it is supported in the [current LTS version of Node (18)](https://nodejs.org/docs/latest-v18.x/api/esm.html)

The feature has been [considered “Stable” since Node 12](https://nodejs.org/docs/latest-v12.x/api/esm.html)

### CODE REVIEW

File: dotcom-rendering/src/lib/contributions.test.ts [hunk 1]
```diff
@@ -1,12 +1,6 @@
 import { setCookie } from '@guardian/libs';
-import {
-	getLastOneOffContributionTimestamp,
-	isRecentOneOffContributor,
-	lazyFetchEmailWithTimeout,
-	ONE_OFF_CONTRIBUTION_DATE_COOKIE,
-	SUPPORT_ONE_OFF_CONTRIBUTION_COOKIE,
-} from './contributions';
-import { getIdApiUserData } from './getIdapiUserData';
+import type { PurchaseInfo } from '@guardian/support-dotcom-components/dist/shared/src/types/targeting';
```
Comment (JamieB-gu) [inline dotcom-rendering/src/lib/contributions.test.ts:2]: Should we update the package to export this type? @tjmw

File: dotcom-rendering/cdk.json [hunk 2]
```diff
@@ -1,5 +1,5 @@
 {
-  "app": "npx ts-node --project=tsconfig.cdk.json cdk/bin/cdk.ts",
+  "app": "npx ts-node --esm --project=tsconfig.cdk.json cdk/bin/cdk.cts",
```
[the remark this answers is not in the document]
Comment (mxdvl) [reply dotcom-rendering/cdk.json:2]: CommonTS 😂

File: dotcom-rendering/scripts/islands/islands.html [hunk 3]
```diff
@@ -74,7 +74,7 @@
 	</style>
 	<h1>Guardian Islands</h1>
 	<h2>Automatically <a
-			href="https://github.com/guardian/dotcom-rendering/blob/66f2501ff3b8be3d8395448b3c3e029ee8bd6b2b/dotcom-rendering/scripts/islands/descriptions.mjs">generated
+			href="https://github.com/guardian/dotcom-rendering/blob/66f2501ff3b8be3d8395448b3c3e029ee8bd6b2b/dotcom-rendering/scripts/islands/descriptions.js">generated
```
[the remark this answers is not in the document]
Comment (mxdvl) [reply dotcom-rendering/scripts/islands/islands.html:77]: Good point, the link was broken anyways…

File: dotcom-rendering/src/lib/braze/hasRequiredConsents.test.ts [hunk 4]
```diff
@@ -1,16 +1,22 @@
...
 let mockOnConsentChangeResult: any;
-jest.mock('@guardian/consent-management-platform', () => ({
-	onConsentChange: (callback: any) => {
-		callback(mockOnConsentChangeResult);
-	},
-	// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
-	getConsentFor: jest.requireActual('@guardian/consent-management-platform')
-		.getConsentFor,
-}));
+jest.mock('@guardian/consent-management-platform', () => {
```
[the remark this answers is not in the document]
Comment (mxdvl) [reply dotcom-rendering/src/lib/braze/hasRequiredConsents.test.ts:6]: It seem like the package also exports ESM, so we can use `unstable_mockModule`, actually.

File: dotcom-rendering/src/lib/messagePicker.test.tsx [hunk 5]
```diff
@@ -1,11 +1,19 @@
...
+const record = jest.fn<OphanRecordFunction>();
+
+beforeAll(() => {
+	record.mockReset();
+	if (window.guardian.ophan) {
+		window.guardian.ophan.record = record;
+	} else {
+		throw new Error('Ophan not on window object. Impossible to test');
+	}
+});
```
[the remark this answers is not in the document]
Comment (mxdvl) [reply dotcom-rendering/src/lib/messagePicker.test.tsx]: No good reason so I’ve reverted to using a mock.

I was battling with the new mocking semantics and got carried away by trying to remove mocks altogether, but this falls way outside the scope of this PR.

File: dotcom-rendering/scripts/webpack/browser-targets.js [hunk 6]
```diff
@@ -3,9 +3,8 @@
  * 1) derive browser targets from @guardian/browserslist-config
  * 2) upgrade any unsupported browsers as described below.
  */
-const getTargetsFromBrowsersList =
-	require('@babel/helper-compilation-targets').default;
```
Comment (mxdvl) [inline dotcom-rendering/scripts/webpack/browser-targets.js:7]: Maybe a createRequire is needed here…

Comment (mxdvl) [reply dotcom-rendering/scripts/webpack/browser-targets.js]: No, it was a matter of leaving this file as CJS.

File: dotcom-rendering/.storybook/main.cjs
Comment (mxdvl) [inline dotcom-rendering/.storybook/main.cjs:1]: This file is well cursed… you can write CommonJS in ESM: even if the extension is `.cts`, you can use static imports and access `__dirname`. Stay well clear.

File: dotcom-rendering/src/lib/contributions.test.ts (hunk 1 above)
Comment (mxdvl) [reply dotcom-rendering/src/lib/contributions.test.ts:2]: It can actually be inferred as is done in https://github.com/guardian/dotcom-rendering/blob/1f1324951172d79c5e29ceabda1a3eece589c4d6/dotcom-rendering/src/lib/contributions.ts#L276

File: dotcom-rendering/cypress.config.js [hunk 7]
```diff
@@ -1,11 +1,14 @@
 import path from 'node:path';
+import { fileURLToPath } from 'node:url';
```
Comment (georgeblahblah) [inline dotcom-rendering/cypress.config.js:2]: I'm not sure if we're using this?

**Review by georgeblahblah — APPROVED** [review]

File: dotcom-rendering/cypress.config.js (hunk 7 above)
Comment (mxdvl) [reply dotcom-rendering/cypress.config.js:2]: Follow-up PR: try to enforce linting and running TS against this file, see https://github.com/guardian/dotcom-rendering/pull/8482/files#diff-ed2946c80569ace27e3818925bb9fbcd193e0292796204b0390f21435b0929a0R22-R24

Good luck!

File: dotcom-rendering/cypress.config.js [hunk 8]
```diff
@@ -34,25 +37,16 @@ module.exports = defineConfig({
...
-					},
+			webpackConfig.webpackOptions.module.rules = [
+				{
+					test: /\.c?tsx?$/,
+					exclude: babelExclude,
+					use: swcLoader,
 				},
-			});
+			];
+
```
Comment (JamieB-gu) [inline dotcom-rendering/cypress.config.js:49]: What's happening here? Is there a change from `ts-loader` to `swcLoader` and is that related to the ESM changes?

File: dotcom-rendering/jest.config.cjs [hunk 9]
```diff
@@ -1,12 +1,25 @@
...
+		'js',
+		'mjs',
+		'cts',
+		'cjs',
+		'jsx',
+		'ts',
+		'tsx',
+		'json',
+		'node',
+	],
```
Comment (JamieB-gu) [inline dotcom-rendering/jest.config.cjs:17]: In other places I think you've referenced a `.ctsx` extension? Is that relevant here?

Comment (mxdvl) [reply dotcom-rendering/jest.config.cjs:17]: I think I only referenced it as a potential match for [`/\.(c|m)?(t|j)sx?/i`](https://regex101.com/r/P6JHq1/1)? I was just trying to cover all bases, even if the file do not actually appear in the codebase…
